MailContainerTest/Services/MailTransferService.cs

```diff
@@ -1,93 +1,77 @@
 ﻿using MailContainerTest.Data;
+using MailContainerTest.Providers;
 using MailContainerTest.Types;
+using Microsoft.Extensions.Configuration;
 using System.Configuration;
 
 namespace MailContainerTest.Services
 {
     public class MailTransferService : IMailTransferService
     {
-        public MakeMailTransferResult MakeMailTransfer(MakeMailTransferRequest request)
+        private IMailContainerStoreProvider _mailContainerStoreProvider;
+        private IMailContainerDataStore _mailContainerDataStore;
+
+        public MailTransferService(IMailContainerStoreProvider mailContainerStoreProvider)
         {
-            var dataStoreType = ConfigurationManager.AppSettings["DataStoreType"];
+            _mailContainerStoreProvider = mailContainerStoreProvider;
+
+            _mailContainerDataStore = _mailContainerStoreProvider.GetDataStore();
+        }
 
-            MailContainer mailContainer = null;
+        public MakeMailTransferResult MakeMailTransfer(MakeMailTransferRequest request)
+        {
+            var result = new MakeMailTransferResult();
 
-            if (dataStoreType == "Backup")
+            try
             {
-                var mailContainerDataStore = new BackupMailContainerDataStore();
-                mailContainer = mailContainerDataStore.GetMailContainer(request.SourceMailContainerNumber);
+                var sourceMailContainer = _mailContainerDataStore?.GetMailContainer(request.SourceMailContainerNumber);
 
-            } else
+                var destinationMailContainer = _mailContainerDataStore?.GetMailContainer(request.DestinationMailContainerNumber);
+
+                if (sourceMailContainer != null
+                    && destinationMailContainer != null
+                    && sourceMailContainer.AllowedMailType.HasFlag(request.MailType)
+                    && CanPerformTransferBetweenContainers(sourceMailContainer, destinationMailContainer))
+                {
+                    var transferSucessful = PerformTransfer(request.NumberOfMailItems, sourceMailContainer, destinationMailContainer);
+
+                    result.Success = transferSucessful;
+                }
+            }
+            catch (Exception ex)
             {
-                var mailContainerDataStore = new MailContainerDataStore();
-                mailContainer = mailContainerDataStore.GetMailContainer(request.SourceMailContainerNumber);
+                // Logging
             }
 
-            var result = new MakeMailTransferResult();
+            return result;
+        }
+
+        private bool CanPerformTransferBetweenContainers(MailContainer sourceContainer, MailContainer destinationContainer)
+        {
```

Comment:
CanPerformTransferBetweenContainers - this returns a bool at the moment. If I had more time I would have returned a more complex object that says why the transfer cannot happen, so this can be logged or returned to the user.

I would also have looked to extract this out into its own class so that the MailTransferServices only job is to make the transfer, rather than worry about eligibility rules.